src/graphql/mirror.test.js

```diff
@@ -3627,6 +3628,58 @@ describe("graphql/mirror", () => {
     });
   });
 
+  describe("_log_request, _log_response, _log_requestUpdateId", () => {
+    it("network requests and responses are saved correctly", async () => {
+      const postQuery = jest.fn();
+      const db = new Database(":memory:");
+      const mirror = new Mirror(db, buildGithubSchema());
+      mirror.registerObject({typename: "Repository", id: "repo:foo/bar"});
+      postQuery.mockImplementationOnce(() =>
+        Promise.resolve({
+          owndata_0: [
+            {
+              __typename: "Repository",
+              id: "repo:foo/bar",
+              url: "url://foo/bar",
+            },
+          ],
+          node_0: {
+            id: "repo:foo/bar",
+            issues: {
+              totalCount: 1,
+              pageInfo: {hasNextPage: false, endCursor: "cursor:repo@1"},
+              nodes: [],
+            },
+          },
+        })
+      );
+
+      await mirror.update(postQuery, {
+        nodesOfTypeLimit: 2,
+        nodesLimit: 3,
+        connectionLimit: 4,
+        connectionPageSize: 5,
+        since: new Date(123),
+        now: () => new Date(),
+      });
+
+      const results = await db
+        .prepare(
+          "SELECT query, query_parameters, update_id, response FROM network_log"
+        )
+        .get();
+
+      expect(results).toMatchSnapshot();
```

Comment:
Snapshot testing is a bit of a brute force implement. It only works for
literal equality checks, and the ease with which one can write snapshot
tests makes it tempting to make tests that are brittle. This in turn
gets folks in the habit of giving snapshots only cursory inspections and
even updating them blindly. When a snapshot test fails, it’s often hard
to tell whether the change is a legitimate error, a flake, or a legacy
value that needs to be updated. Instead of snapshot tests, a targeted
test that checks _properties_ of the relevant value is often clearer and
more useful.

Thus, we try to use snapshot tests only for things that are really
unwieldy and large, *and* that can’t reasonably be reduced to something
more targeted. Note that the module that you’re currently editing has
3500 lines of tests, and yet only two snapshot tests among them, and
those two snapshots are explicitly for integration tests that are meant
to be copy-pastable from the snapshot test file the GitHub GraphQL UI
(as documented in the comment on the test case).

This newly added snapshot test, for instance, would spuriously fail if
the system under test changed the query name `MirrorUpdate` to `MU`.
(e.g., as a refactor to reduce network bandwidth).

Please replace these snapshot tests with something more precise. It may
be helpful to follow the pattern used by the rest of the tests: we test
the helper methods carefully and precisely, and then simply check that
`update` invokes the helper methods correctly. If you follow this
pattern, then you can control exactly what gets passed to `_logRequest`
and friends. This will also give you the opportunity to test that the
query variables are serialized correctly even when non-empty. You can
update the existing test, “update > delegates and updates the database”
to add the checks that the helpers are being invoked correctly.

Reply:
Thanks again for a great explanation! Updated.